flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/SnapshotManager.java

```diff
@@ -90,6 +90,22 @@ public boolean snapshotExists(long snapshotId) {
         }
     }
 
+    public @Nullable Long earlierThanTimeMills(long timestampMills) {
+        Long earliest = earliestSnapshotId();
+        Long latest = latestSnapshotId();
+        if (earliest == null || latest == null) {
+            return null;
+        }
+
+        for (long i = latest; i >= earliest; i--) {
+            long commitTime = snapshot(i).timeMillis();
+            if (commitTime < timestampMills) {
+                return i;
+            }
+        }
+        return earliest - 1;
```

Comment:
Why we return `earliest -1` here instead of `earliest`？

Reply:
`earlierThanTimeMills`.
Here should return previous snapshot for this time mills.
And then use this `earlierSnapshot + 1` to read incremental data.

Reply:
Nice, we can doc it, otherwise it may cause misunderstanding by other ones

Reply:
The earliest snapshots here refer to those that have not expired.
earliest should be changed after expiration.